### PR TITLE
GODRIVER-4118 fix: correct Truncate boundary handling for multi-byte UTF-8 characters

### PR DESCRIPTION
**Pull request by zigzagdev**
[GODRIVER-4118](https://jira.mongodb.org/browse/GODRIVER-4118)

## Summary

Fix `Truncate` in `internal/bsoncoreutil` to correctly handle the case where `width` falls exactly on a UTF-8 multi-byte character boundary.

## Background & Motivation

The previous implementation of `Truncate` missed the case where `width` falls exactly at the end of a multi-byte UTF-8 character.

F.E

```go
Truncate("a©bc", 3) // © is encoded as 2 bytes: 0xC2 0xA9
```

`str[:3]` ends cleanly after `©`, but the old logic checked the last byte (`0xA9`, pattern `10xxxxxx`) and incorrectly treated it as a dangling continuation byte, stepping back past the complete `©` and returning `"a"` instead of the correct `"a©"`.

The same bug affected the expected values in `document_test.go` and `value_test.go`, which were written against the incorrect behaviour.

## What I have done

- Rewrote `Truncate` using `unicode/utf8`:

1. Step back over trailing continuation bytes (`10xxxxxx`) to locate the start of the last rune within the `width`-byte prefix.
2. Decode that rune with `utf8.DecodeRuneInString` to get its true byte length.
3. If `start + size > width` the rune was cut off and is dropped (`str[:start]`); otherwise the prefix is already valid and `str[:width]` is returned as-is.

## Test Plans

- Added two regression test cases to `bsoncoreutil_test.go` that fail against the old implementation and pass with the fix:

| case | input | width | old result | expected |
|---|---|---|---|---|
| `multi-byte exact boundary` | `"a©bc"` | 3 | `"a"` | `"a©"` |
| `multi-byte exact boundary before trailing ascii` | `"你好!"` | 6 | `"你"` | `"你好"` |

Both cases cover `width` landing exactly on the last byte of a multi-byte character, which is the scenario the old code mishandled.

### CODE REVIEW

**Comment by matthewdale** [comment]
@zigzagdev Thanks for the fix! I've created [GODRIVER-4118](https://jira.mongodb.org/browse/GODRIVER-4118) to track the review. We'll prioritize that ticket at our next planning meeting next week.

**Comment by zigzagdev** [comment]
@matthewdale 

Thanks for setting up the ticket! 
I'll keep an eye on [GODRIVER-4118](https://jira.mongodb.org/browse/GODRIVER-4118). 

Happy to make any changes if the review turns up anything.

File: internal/bsoncoreutil/bsoncoreutil.go [hunk 1]
```diff
@@ -16,25 +20,24 @@ func Truncate(str string, width int) string {
...
-		for i := len(newStr) - 1; i >= 0; i-- {
-			if newStr[i]&0xC0 == 0xC0 {
-				// Truncate at the end of the character before the character we stepped back to
-				return newStr[:i]
-			}
-		}
+	// Decode the rune starting at start from the original (untruncated)
+	// string to determine its true byte length. If that rune extends past
+	// width, it was cut off and must be dropped entirely.
+	_, size := utf8.DecodeRuneInString(str[start:])
```
Comment (prestonvasquez) [inline internal/bsoncoreutil/bsoncoreutil.go:37]: [blocking] DecodeRuneInString creates a performance regression in truncate, use [RuneStart ](https://pkg.go.dev/unicode/utf8#RuneStart) instead.

Additionally, we should add benchmarking to prove the performance improvement: https://gist.github.com/prestonvasquez/f1a0c1ce7e641256f86ad9bf7b546335

File: internal/bsoncoreutil/bsoncoreutil_test.go [hunk 2]
```diff
@@ -45,6 +45,18 @@ func TestTruncate(t *testing.T) {
...
+			arg:      "a©bc",
+			width:    3,
+			expected: "a©",
+		},
+		{
+			name:     "multi-byte exact boundary before trailing ascii",
+			arg:      "你好!",
+			width:    6,
+			expected: "你好",
+		},
```
Comment (prestonvasquez) [inline internal/bsoncoreutil/bsoncoreutil_test.go:59]: [blocking] We should also add tests for a negative width, invalid UTF-8, and 4 byte runes.

**Review by prestonvasquez — CHANGES_REQUESTED** [review]

File: internal/bsoncoreutil/bsoncoreutil_test.go (hunk 2 above)
Comment (zigzagdev) [reply internal/bsoncoreutil/bsoncoreutil_test.go:60]: @prestonvasquez 

[fixed](https://github.com/mongodb/mongo-go-driver/pull/2564/commits/3d788a25222172334bf57723a632182f261e5893) it.

Updated for covering negative width, 4-byte runes (boundary + split), and invalid UTF-8.

File: internal/bsoncoreutil/bsoncoreutil.go (hunk 1 above)
Comment (zigzagdev) [reply internal/bsoncoreutil/bsoncoreutil.go]: @prestonvasquez 

[fixed](https://github.com/mongodb/mongo-go-driver/pull/2564/commits/3ec59248e17eb5f8b0de2014a2793f03e80ff3f2) it.

Switched to `utf8.RuneStart` on the byte after the cut point instead of decoding.

File: internal/bsoncoreutil/bsoncoreutil_test.go [hunk 3]
```diff
@@ -56,3 +105,56 @@ func TestTruncate(t *testing.T) {
...
+		width int
+	}{
+		// Width lands mid-string on an ASCII byte, so utf8.RuneStart reports a
+		// clean cut immediately and the backward scan never runs. This is the
+		// common logging case: truncating extended JSON.
+		{"ascii_mid_string", ascii, benchWidth},
+
+		// 999 == 333*3, so width lands exactly on the end of a 3-byte rune and
+		// the full prefix is kept. This is the path PR #2564 corrected. Scans
+		// back over 2 continuation bytes.
```
Comment (prestonvasquez) [inline internal/bsoncoreutil/bsoncoreutil_test.go:142]: [question] This case shouldn't scan at all. Are you seeing that behavior?

File: internal/bsoncoreutil/bsoncoreutil_test.go [hunk 4]
```diff
@@ -56,3 +105,56 @@ func TestTruncate(t *testing.T) {
...
+		{"exact_rune_boundary", threeByte, 999},
+
+		// 999 == 249*4+3, so width lands 3 bytes into a 4-byte rune, which is
+		// dropped entirely. Scans back over 2 continuation bytes.
```
Comment (prestonvasquez) [inline internal/bsoncoreutil/bsoncoreutil_test.go:146]: [question] This case should scan 3 times, not 2. Are you seeing that behavior?
